Introduce ManualRoutingParameters

shaavan opened the pull request:
Partially addresses #3262 

- Introduce a new struct `ManualRoutingParameters` that optionally takes the parameter to set for routing.
- Pass it to `pay_for_offer` to use it along with the invoice's payment parameter to set route parameters.

shaavan commented:
Hey @TheBlueMatt @tnull,

A gentle ping! Would love to get your feedback or approach ACK before moving forward with testing and applying a similar approach on the BOLT11 side.

Thanks a lot!

vincenzopalazzo commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +962,43 @@ impl PaymentParameters {
+
+impl ManualRoutingParameters {
+	/// Initates an empty set of [`ManualRoutingParameters`]
+	pub fn new() -> Self {
```
Random though: We would like to use an `fn empty` name for this and keep the `new` for a common constructor (that I do not know what common in this case means 😄 )

shaavan replied:
That makes sense!

I noticed, in the codebase, we're using `fn new` more often for blank initialization, though, so to keep things consistent, it might be a good idea to stick with `new`. Lemmino what you think! :)

vincenzopalazzo commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2238,7 +2244,7 @@ impl_writeable_tlv_based_enum_upgradable!(PendingOutboundPayment,
 	(5, AwaitingInvoice) => {
 		(0, expiration, required),
 		(2, retry_strategy, required),
-		(4, max_total_routing_fee_msat, option),
```
Is this breaking backward compatibility?

Matt taught me how this works a few weeks ago, but I’m not completely sure if I’m understanding it correctly, so please bear with me.

The old LDK node has stored a `u64` with the type `4`. So, if you now change the format, there will be a type mismatch between what LDK 124 has stored and what the future LDK is expecting.

If I’m understanding this correctly, you need a new type (a new number) for `manual_routing_params`. When upgrading, the `max_total_routing_fee_msat` will have a value, and from that, you can create the `manual_routing_params`. Then, you will store `manual_routing_params` as `Some` or `None`, depending on the value.

shaavan replied:
Thanks a lot for the pointer, and guidance!
Reintroduced `max_total_routing_fee_msat` in [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03) to support downgrading.

vincenzopalazzo commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9303,7 +9309,7 @@ where
 		payer_note: Option<String>, payment_id: PaymentId, retry_strategy: Retry,
-		max_total_routing_fee_msat: Option<u64>
+		manual_routing_params: Option<ManualRoutingParameters>
 	) -> Result<(), Bolt12SemanticError> {
```
OK no needed option here but: what do you think about having the function call like `pay_for_offer(offer, OfferParams::default())` where `OfferParams` is something like that

```rust
struct OfferParams {
    quantity: Option<u64>,
    amount_msats: Option<u64>,
    manual_routing_params: Option<ManualRoutingParameters>,
}
```

I think is  a lot of more verbose in this way, but I also this that this will be a lot of cleaner.

Now I do not think this will fit well in this PR maybe can be a followup one, but with this patter, it is possible to hide future updates to offers (e.g: recurrence or market place feature) without breaking too much the code

shaavan replied:
I think this is a great improvement! It will help keep the function signature cleaner while remaining maintainable for future parameter increases.

Maybe we can take this in a follow-up. @thebluematt, what do you think?

vincenzopalazzo reviewed (commented):
LGTM overall.

However I am not the right person to judge if the PR is what the ldk team is looking for.

For instance the following is confusing me :) but I think this is a discussion to be done inside the issue

>They also don't really make sense in a BOLT 11 world if we pass a Bolt11Invoice directly to ChannelManager (which we should do now that lightning depends on lightning-invoice).

BTW Nice git history!

P.S: I left some API idea comment in some commit review, idk if they are useful or not. probably they will remove a lot of `map(_ {})` code, but I had to write the code to confirm

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +980,43 @@ impl PaymentParameters {
 
+/// Information manually provided by the user to route the payment
+#[derive(Clone, Copy)]
+pub struct ManualRoutingParameters {
```
How about `RouteParameterOverrides`, `RouteParametersConfig`, or `CustomizedRouteParameters`? Most fields are for overriding `PaymentParameters`, but that is a part of `RouteParameters` one of which can be overridden, too.

shaavan replied:
Updated the name to `RouteParametersOverride` in [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03) as it sounded to me the most optimum for now.

Thanks, @jkczyz for the suggestion on names!

@TheBlueMatt
Let me know if the name looks good to you! Thanks!

shaavan replied:
Updated the name to `RouteParametersConfig` in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04), to represent its role as allowing users to configure route parameters.
Thanks for the name suggestions! 🌟

shaavan commented:
Updated from [pr3342.01](https://github.com/shaavan/rust-lightning/commits/pr3342.01) to [pr3342.02](https://github.com/shaavan/rust-lightning/commits/pr3342.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.01..pr3342.02)):

Changes:
1. Rebase on main

shaavan commented:
Updated from [pr3342.02](https://github.com/shaavan/rust-lightning/commits/pr3342.02) to [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.02..pr3342.03)):
Addressed @vincenzopalazzo, @jkczyz, @TheBlueMatt comments

Changes:
1. Renamed `ManualRoutingParameters` -> `RouteParametersOverride`
2. Reintroduce the `max_total_routing_fee_msat` in `AwaitingInvoice` to support downgrading.
3. Update API, so that the `PaymentParameters` are created using the Overrides, and then the `RouteParameters` are created using it.

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -849,10 +849,12 @@ impl OutboundPayments {
-					max_total_routing_fee_msat = *max_total_fee;
+					max_total_routing_fee_msat = manual_routing_params.map(
+						|params| params.max_total_routing_fee_msat
+					).flatten();
```
Done! [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03)
Thanks for the pointer!

shaavan commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +980,43 @@ impl PaymentParameters {
+	/// Same as [`PaymentParameters::max_path_count`]
+	pub max_path_count: Option<u8>,
+	/// Same as [`PaymentParameters::max_channel_saturation_power_of_half`]
+	pub max_channel_saturation_power_of_half: Option<u8>,
```
I’ve made all the fields optional, so they act as overrides. If left as `None`, the values from `PaymentParameters` (created using the invoice and default settings) will be used. I’ve also updated the docs to make this clearer for this struct.

Would love to hear your feedback on this API design! Thanks so much!

shaavan replied:
Expanded the role of struct in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04), allowing configuring parameters to the user instead of simply overriding them.
Hence remove the unnecessary option, and use default if the value is absent.
Thanks for the pointer!

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -871,12 +871,10 @@ impl OutboundPayments {
 
-		let mut route_params = RouteParameters::from_payment_params_and_value(
-			PaymentParameters::from_bolt12_invoice(&invoice), invoice.amount_msats()
+		let route_params = RouteParameters::from_payment_and_manual_params(
```
Update the API design in [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03)
Thanks a lot for the suggestion! 🌟

shaavan commented:
Updated from [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03) to [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.03..pr3342.04)):
Addressed @TheBlueMatt’s comments:

Changes:

### Update Role of RouteParamsOverride:
1. Instead of presenting this new struct as just overriding the PaymentParams, it is now structured as allowing the user to set configuration parameters.
2. This brings some changes to the struct:
    - It has been renamed from `RouteParamsOverride` to `RouteParamsConfig`.
    - The parameters are now required rather than optional.
    - Documentation has been expanded to make the struct stand independently of `{Route, Payment} Params`.

### Cleanups:
1. Introduced a new separate function, `with_user_config`, in `PaymentParams`, making the function more modular and focused.
2. `RouteParamsConfig` is now a required field in `AwaitingInvoice` and `InvoiceReceived`, initialized with default values if absent. This eliminates redundant optioning and switches to default values where necessary.

shaavan commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +970,46 @@ impl PaymentParameters {
 	}
 }
 
+/// A struct for overriding specific [`RouteParameters`] values created
```
Thanks for the pointer, Matt!
If RouteParameters will not be seen by the user after the update, it makes sense for the struct to stand independent of it.
I have updated the struct be positioned as allowing the user to specify configuration parameters, and updated the documentation to reflect that in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04)
Thanks a lot for suggestion!

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2240,11 +2255,12 @@ impl_writeable_tlv_based_enum_upgradable!(PendingOutboundPayment,
 	(7, InvoiceReceived) => {
 		(0, payment_hash, required),
 		(2, retry_strategy, required),
-		(4, max_total_routing_fee_msat, option),
```
Ah right! Looks like we have to keep it for downgrade support! Missed opportunity for cleanup 🥲
Reverted in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04) Thanks for the pointer! 🌟

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -61,7 +61,11 @@ pub(crate) enum PendingOutboundPayment {
+		// Deprecated: Retained for backward compatibility.
+		// If set during read, this field overrides `RouteParameters::max_total_routing_fee_msat`
+		// instead of `RouteParametersOverride::max_total_routing_fee_msat`.
 		max_total_routing_fee_msat: Option<u64>,
```
Done! 🚀 Thanks a lot!

shaavan commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +980,43 @@ impl PaymentParameters {
+	/// Same as [`RouteParameters::max_total_routing_fee_msat`]
+	pub max_total_routing_fee_msat: Option<u64>,
+	/// Same as [`PaymentParameters::max_total_cltv_expiry_delta`]
+	pub max_total_cltv_expiry_delta: Option<u32>,
```
Fixed in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04)
Thanks a lot!

shaavan commented:
Updated from [pr3342.03](https://github.com/shaavan/rust-lightning/commits/pr3342.03) to [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.03..pr3342.04)):
Addressed @TheBlueMatt’s comments:

Changes:

### Update Role of RouteParamsOverride:
1. Instead of presenting this new struct as just overriding the PaymentParams, it is now structured as allowing the user to set configuration parameters.
2. This brings some changes to the struct:
    - It has been renamed from `RouteParamsOverride` to `RouteParamsConfig`.
    - The parameters are now required rather than optional.
    - Documentation has been expanded to make the struct stand independently of `{Route, Payment} Params`.

### Cleanups:
1. Introduced a new separate function, `with_user_config`, in `PaymentParams`, making the function more modular and focused.
2. `RouteParamsConfig` is now a required field in `AwaitingInvoice` and `InvoiceReceived`, initialized with default values if absent. This eliminates redundant optioning and switches to default values where necessary.

shaavan commented on lightning/src/routing/router.rs:
```diff
@@ -962,6 +980,43 @@ impl PaymentParameters {
+#[derive(Clone, Copy)]
+pub struct ManualRoutingParameters {
+	/// Same as [`RouteParameters::max_total_routing_fee_msat`]
+	pub max_total_routing_fee_msat: Option<u64>,
```
Expanded the docs, explaining the default behavior in [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04)
Thanks a lot for the pointer!

shaavan commented:
@TheBlueMatt

> Looks like most of my previous comments went unaddressed? Just checking cause I interpreted your comment as you addressing them.

Hey Matt, I’m really sorry for missing your previous comments – not sure how that slipped through! I've gone through everything now and made sure to address them all in the latest update. Thanks so much for your patience! 🌟

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -849,14 +856,18 @@ impl OutboundPayments {
 					retry_strategy = *retry;
-					max_total_routing_fee_msat = *max_total_fee;
+					// If max_total_fee is present, update route_params_config with the specified fee.
+					// This supports the standard behavior during downgrades.					
```
That's the behavior needed when upgrading, not downgrading. For downgrades, we just need to make sure we continue to write `max_total_routing_fee_msat`.

shaavan replied:
Makes sense! Looks like I mixed up supporting downgrade with upgrade :)
Thanks for the pointer!
Fixed in [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06)

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1608,7 +1624,9 @@ impl OutboundPayments {
 					retry_strategy,
-					max_total_routing_fee_msat,
+					route_params_config,
+					// Retained for downgrade support.
+					max_total_routing_fee_msat: None,
```
If you set this to `None`, then downgrades will not retain the value since they don't know how to read `route_params_config`. We need to set both in order to support downgrades.

shaavan replied:
Ah, right! I was not writing that so that we can phase it out in future versions because it will be unused. But now I understand that we would still need to write it so that older versions can find and read this value.
Thanks a lot for the pointer!
Fixed in [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06)

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1031,7 +1042,7 @@ impl OutboundPayments {
 					let pay_params = PaymentParameters::from_static_invoice(invoice);
 					let mut route_params = RouteParameters::from_payment_params_and_value(pay_params, amount_msat);
-					route_params.max_total_routing_fee_msat = *max_total_routing_fee_msat;
+					route_params.max_total_routing_fee_msat = route_params_config.map(|params| params.max_total_routing_fee_msat).flatten();
```
This doesn't compile since `route_param_config` is not an option. FYI, use the following to make sure the compiler sees this code: `!RUSTFLAGS="--cfg=async_payments" cargo +1.63.0 check -p lightning`.

shaavan replied:
Thanks a lot for the suggestion on the code! It will be super helpful for future references too!

I also noticed that the function was not using the rest of params_config when building PaymentParams. I have also fixed this bug in [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06)

Thanks for the pointer!

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1607,14 +1607,10 @@ impl OutboundPayments {
-			|fee_msats| RouteParametersConfig::new()
-				.with_max_total_routing_fee_msat(fee_msats)
-		);
+		let route_params_config = route_params_config.unwrap_or(RouteParametersConfig::new());
```
In this commit, we should also set `max_total_routing_fee_msat` in `AwaitingInvoice` to what is in the config in order to support downgrades.

shaavan replied:
Done! Update the code in [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06) to write max_total_routing_fee_msat to support downgrades too!
Thanks for the suggestion!

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -883,11 +886,11 @@ impl OutboundPayments {
 		}
 
 		let mut route_params = RouteParameters::from_payment_params_and_value(
-			PaymentParameters::from_bolt12_invoice(&invoice), invoice.amount_msats()
+			PaymentParameters::from_bolt12_invoice(&invoice).with_user_config(params_config), invoice.amount_msats()
 		);
-		if let Some(max_fee_msat) = max_total_routing_fee_msat {
-			route_params.max_total_routing_fee_msat = Some(max_fee_msat);
-		}
+
+		params_config.max_total_routing_fee_msat.map(|fee_msat| route_params.max_total_routing_fee_msat = Some(fee_msat));
```
Prefer `if let` instead of putting a side effect in `map`.

shaavan replied:
Done! [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06)
The code looks much more readable now!

Thanks for the pointer!

shaavan commented:
Updated from [pr3342.04](https://github.com/shaavan/rust-lightning/commits/pr3342.04) to [pr3342.05](https://github.com/shaavan/rust-lightning/commits/pr3342.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.04..pr3342.05)):

Changes:
1. Rebase on main. Rest no changes.

shaavan commented:
Updated from [pr3342.05](https://github.com/shaavan/rust-lightning/commits/pr3342.05) to [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.05..pr3342.06)):
Addressed @jkczyz, @TheBlueMatt comments

Changes:
1. Fix the comment regarding supporting standard behaviour during upgrade.
2. Set the max_total_routing_fee_msat along with route_params config to support downgrade.
3. Fix the `fn static_invoice_received` to ensure proper compilation and behaviour `async_payment` case as well.
4. Refactor the code to use if let instead of map for better readability.
5. Introduce the full config parameter for create_refund_builder as well.
6. Structured the commits slightly to have better logical progression.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9208,10 +9210,15 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
 
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop($self);
 
+		let route_params_config = max_total_routing_fee_msat.map(
```
Yep, I believe we should!
The reason I held off until now was that we were still in the testing phase, and I wanted to make sure we nailed down a solid approach. I also wanted to ensure I could get a basic, working version of the branch ready before diving into the refund code.

Now that the version looks solid, I’ve gone ahead and introduced the params config as a parameter for the refund as well!

Thanks so much for the nudge!

shaavan commented:
> Okay, we probably shouldn't have held this up on https://github.com/lightningdevkit/rust-lightning/pull/3378 now that it took way too long to land, but it did finally land, care to rebase on that?

I am glad, it finally landed! Rebase on the way! 🚂

shaavan commented:
Updated from [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06) to [pr3342.07](https://github.com/shaavan/rust-lightning/commits/pr3342.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.06..pr3342.07)):
Addressed @TheBlueMatt comments:

Changes:
1. Rebase on main.

shaavan commented:
Updated from [pr3342.06](https://github.com/shaavan/rust-lightning/commits/pr3342.06) to [pr3342.07](https://github.com/shaavan/rust-lightning/commits/pr3342.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.06..pr3342.07)):
Addressed @TheBlueMatt comments:

Changes:
1. Rebase on main.

shaavan commented:
@TheBlueMatt

Hi, after rebasing, I tried using the new syntax to handle the legacy fields. Here's the diff I wrote:  

```diff
(5, AwaitingInvoice) => {
                (0, expiration, required),
                (2, retry_strategy, required),
-               (4, max_total_routing_fee_msat, option),
+               (4, _max_total_routing_fee_msat, (legacy, u64,
+                       |us: &PendingOutboundPayment| match us {
+                               PendingOutboundPayment::AwaitingInvoice { route_params_config, .. } => Some(route_params_config.max_total_routing_fee_msat),
+                               _ => None,
+                       }
+               )),
                (5, retryable_invoice_request, option),
-               (7, route_params_config, (default_value, RouteParametersConfig::new())),
+               (7, route_params_config, (default_value, (
+                       if let Some(fee_msat) = _max_total_routing_fee_msat {
+                               RouteParametersConfig::new().with_max_total_routing_fee_msat(fee_msat)
+                       } else {
+                               RouteParametersConfig::new()
+                       }
+               ))),
        },
```

The code compiles, but I ran into errors in multiple tests. I would love to get your insight on this—am I missing something or approaching it incorrectly? Thanks a lot!

jkczyz commented:
> ```diff
> with_max_total_routing_fee_msat
> ```

Discussed offline. Just needed to remove the `Some` from the legacy write.  Also, wondering if it should be `Option<u64>` and if the compiler would then catch it?

jkczyz commented:
> Hmm, that's weird, legacy writes are supposed to be `Option`s. Can you push the code somewhere?

Problem was `max_total_routing_fee_msat` is also an `Option`, so it was being double wrapped. I think we can force the compiler to catch this with an explicit type hint at:

https://github.com/shaavan/rust-lightning/blob/737867c3d44fb0f9922a25827cd654c9247102af/lightning/src/util/ser_macros.rs#L43

Something like:

```rust
{ let value: Option<$fieldty> = $write($($self)?); value }
```

shaavan commented:
> legacy writes are supposed to be Options. 

Yep! I believe that's working perfectly.
The issue was `route_params_config.max_total_routing_fee_msat` is already `Option<u64>`, but I was trying to return `Some(route_params_config.max_total_routing_fee_msat)` which caused the issue.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9881,7 +9884,7 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
 	pub fn create_refund_builder(
 		&$self, amount_msats: u64, absolute_expiry: Duration, payment_id: PaymentId,
-		retry_strategy: Retry, max_total_routing_fee_msat: Option<u64>
+		retry_strategy: Retry, route_params_config: Option<RouteParametersConfig>
```
Given the default `RouteParametersConfig` is used if `None`, perhaps we should remove the `Option` from here and throughout? Drawback is that user call sites already passing `None` would need to be updated on upgrade. @TheBlueMatt not sure if you have a preference here.

shaavan replied:
Thanks a lot for the pointer! My initial reasoning for using `Option` was:

1. It keeps the call site clean.  
2. Since users aren’t manually passing a route parameters config, I felt it was more intuitive to pass `None` rather than `RouteParametersConfig::new()`.

Would love to hear your and Matt’s thoughts on this!

shaavan replied:
Updated the call-site in [pr3342.10](https://github.com/shaavan/rust-lightning/commits/pr3342.10), so now the user calls RouteParametersConfig::default() instead of None in default case
Thanks for the pointer!

shaavan commented:
Updated from [pr3342.08](https://github.com/shaavan/rust-lightning/commits/pr3342.08) to [pr3342.09](https://github.com/shaavan/rust-lightning/commits/pr3342.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.08..pr3342.09)):
Addressed @jkczyz comment

Changes:
1. Updated `_encode_tlv` macro to introduce compile time type check.

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2400,15 +2400,35 @@ impl_writeable_tlv_based_enum_upgradable!(PendingOutboundPayment,
-		// TODO: Add support for legacy `max_total_routing_fee_msat`
+		(4, _max_total_routing_fee_msat, (legacy, u64,
+			|us: &PendingOutboundPayment| match us {
+				PendingOutboundPayment::AwaitingInvoice { route_params_config, .. } => route_params_config.max_total_routing_fee_msat,
```
Thanks a lot for the pointer!
I incorporated this in [pr3342.09](https://github.com/shaavan/rust-lightning/commits/pr3342.09) and also verified that compilation fail if the type doesn't exactly match.

shaavan commented:
Updated from [pr3342.09](https://github.com/shaavan/rust-lightning/commits/pr3342.09) to [pr3342.10](https://github.com/shaavan/rust-lightning/commits/pr3342.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.09..pr3342.10)):
Addressed @jkczyz, @TheBlueMatt comments

Changes:
1. Update call-site so that user can call `RouteParameters::default()` instead of `None`, in default case.
2. Reorder tlv ordering so that legacy is appropriately read before the new field.

shaavan commented:
Updated from [pr3342.10](https://github.com/shaavan/rust-lightning/commits/pr3342.10) to [pr3342.11](https://github.com/shaavan/rust-lightning/commits/pr3342.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.10..pr3342.11)):

Changes:
1. Rebase on main.

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2390,13 +2393,35 @@ impl_writeable_tlv_based_enum_upgradable!(PendingOutboundPayment,
 		(0, payment_hash, required),
 		(2, retry_strategy, required),
-		(4, max_total_routing_fee_msat, option),
+		(3, route_params_config, (default_value, (
```
Thanks, Matt! Since all the tests were passing, I assumed I had used the macro correctly here.  
But after going through the logic, I see that we should be first reading the legacy variable, which means it should have the smaller type compared to the new one.  
I've corrected the code in [pr3342.10](https://github.com/shaavan/rust-lightning/commits/pr3342.10).  

Really appreciate the pointer!

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1076,7 +1076,7 @@ impl_writeable_tlv_based!(RouteParametersConfig, {
 impl RouteParametersConfig {
 	/// Initates an new set of route parameter configs with default parameters.
-	pub fn new() -> Self {
+	pub fn default() -> Self {
```
May as well have this implement the `Default` trait instead.

shaavan replied:
Thanks for the pointer!
Addressed in [pr3342.12](https://github.com/shaavan/rust-lightning/commits/pr3342.12)

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -943,6 +943,16 @@ impl PaymentParameters {
+			max_total_cltv_expiry_delta: params_config.max_total_cltv_expiry_delta,
+			max_path_count: params_config.max_path_count,
+			max_channel_saturation_power_of_half: params_config.max_channel_saturation_power_of_half,
+			.. self
```
nit: remove space before `self`

jkczyz commented:
LGTM. For the first commit's message, you can remove the "f: " prefix since it isn't a fixup (i.e., won't be squashed). @TheBlueMatt I'm happy to have the fixups squashed if you are.

shaavan commented:
Updated from [pr3342.12](https://github.com/shaavan/rust-lightning/commits/pr3342.12) to [pr3342.13](https://github.com/shaavan/rust-lightning/commits/pr3342.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3342.12..pr3342.13)):
Addressed @TheBlueMatt comments:

Changes:
1. Update the documentation and naming for `with_user_config` to be clearer and explicitly state that we ignore the `max_total_routing_fee_msat` application within it.
2. Squash, and clean-up commit history

shaavan commented on lightning/src/routing/router.rs:
```diff
@@ -943,6 +943,16 @@ impl PaymentParameters {
 		}
 	}
 
+	/// Update the parameters with the configuration provided by user.
```
Thanks for the pointer! Addressed in [pr3342.13](https://github.com/shaavan/rust-lightning/commits/pr3342.13)
I updated the naming and change the function to pub(crate) to make it explicit that `max_total_routing_fee_msat` is not applied here, and prevent potential footgunning.
Also, I added an explicit line of documentation to make it clearer.
Let me know if it looks good now! Thanks!
